### PR TITLE
Support dynamic properties in groupby and aggregate (port to 6.x)

### PR DESCRIPTION
**Pull request by kosinsky**
### Issues
*This pull request fixes issue https://github.com/OData/odata.net/issues/760 (port to WebAPI 6.x).*  

### Description
Dynamic properties don't have type. Relaxed null check in AggregateExpression and related changes in ApplyBinder to allow usage of dynamic properties in groupby clause

### Checklist (Uncheck if it is not completed)
- [ x ] Test cases added
- [ x ] Build and test with one-click build and test script passed

### Additional work necessary
Depends on ODL nuget package that will include https://github.com/OData/odata.net/pull/808 and https://github.com/OData/odata.net/pull/816

### CODE REVIEW

File: src/System.Web.OData/OData/Query/Expressions/FilterBinder.cs [hunk 1]
```diff
@@ -301,25 +299,25 @@ private Expression BindPropertyAccessExpression(SingleValueOpenPropertyAccessNod
...
-                edmStructuredType = edmTypeReference.AsComplex().ComplexDefinition();
-            }
-            else
-            {
-                throw Error.NotSupported(SRResources.QueryNodeBindingNotSupported, openNode.Kind, typeof(FilterBinder).Name);
-            }
-            var prop = EdmLibHelpers.GetDynamicPropertyDictionary(edmStructuredType, Model);
-            return prop;
-        }
+        //private PropertyInfo GetDynamicPropertyContainer(SingleValueOpenPropertyAccessNode openNode)
```
Comment (robward-ms) [inline src/System.Web.OData/OData/Query/Expressions/FilterBinder.cs]: Why not delete this?

File: src/System.Web.OData/packages.config [hunk 2]
```diff
@@ -4,8 +4,8 @@
   <package id="Microsoft.AspNet.WebApi.Core" version="5.2.2" targetFramework="net45" />
   <package id="Microsoft.Extensions.DependencyInjection" version="1.0.0" targetFramework="net45" />
   <package id="Microsoft.Extensions.DependencyInjection.Abstractions" version="1.0.0" targetFramework="net45" />
-  <package id="Microsoft.OData.Core" version="7.2.0" targetFramework="net45" />
+  <!--<package id="Microsoft.OData.Core" version="7.2.0" targetFramework="net45" />
```
Comment (robward-ms) [inline src/System.Web.OData/packages.config]: Is this change intentional?

**Comment by robward-ms** [comment]
@kosinsky  - This and the V4 5.x version of the PR reference #760, which is an issue about release notes. Can you reference the correct bug?

I suspect you meant ODL 760, in which case maybe just remove the reference to 760 since GitHub assumes the repo. FYI: https://github.com/OData/odata.net/issues/760

File: src/System.Web.OData/OData/Query/Expressions/FilterBinder.cs (hunk 1 above)
Comment (kosinsky) [reply src/System.Web.OData/OData/Query/Expressions/FilterBinder.cs]: #resolved

File: src/System.Web.OData/packages.config (hunk 2 above)
Comment (kosinsky) [reply src/System.Web.OData/packages.config]: I've rolled that back. However, this PR relies on changes in odata.net. Can I use nightly build for 7.4 here?

---
In reply to: [137667056](https://github.com/OData/WebApi/pull/1004#discussion_r137667056) [](ancestors = 137667056)

**Comment by kosinsky** [comment]
Yes, I meant 760 from ODL.

---
In reply to: [327939928](https://github.com/OData/WebApi/pull/1004#issuecomment-327939928) [](ancestors = 327939928)

File: src/System.Web.OData/OData/ExpressionHelperMethods.cs [hunk 3]
```diff
@@ -248,6 +251,33 @@ public static MethodInfo EnumerableCountGeneric
...
+        }
+
+        public static decimal? SafeConvertToDecimal(object value)
+        {
+            if (value == null || value == DBNull.Value)
+            {
+                return null;
+            }
+
+            var type = value.GetType();
```
Comment (mtroth-microsoft) [inline src/System.Web.OData/OData/ExpressionHelperMethods.cs]: >var typ [](start = 12, length = 7)

nit: var

File: src/System.Web.OData/OData/Query/Expressions/AggregationBinder.cs [hunk 4]
```diff
@@ -388,6 +412,44 @@ private Expression CreatePropertyAccessExpression(Expression source, IEdmPropert
             }
         }
 
+        private Expression CreateOpenPropertyAccessExpression(SingleValueOpenPropertyAccessNode openNode)
+        {
+            var sourceAccessor = BindAccessor(openNode.Source);
```
Comment (mtroth-microsoft) [inline src/System.Web.OData/OData/Query/Expressions/AggregationBinder.cs]: >var s [](start = 12, length = 5)

var hides the actual type when using tools like code flow or notepad.

File: src/System.Web.OData/OData/Query/Expressions/AggregationBinder.cs [hunk 5]
```diff
@@ -447,5 +509,15 @@ private List<NamedPropertyExpression> CreateGroupByMemberAssignments(IEnumerable
 
             return properties;
         }
+
+        private static Expression WrapDynamicCastIfNeeded(Expression propertyAccessor)
```
Comment (mtroth-microsoft) [inline src/System.Web.OData/OData/Query/Expressions/AggregationBinder.cs]: >rivate static Ex [](start = 9, length = 16)

not sure why we've turned off the style cop rule requiring static methods before instance.

File: src/System.Web.OData/OData/Query/Expressions/ExpressionBinderBase.cs [hunk 6]
```diff
@@ -499,6 +500,31 @@ internal string GetFullPropertyPath(SingleValueNode node)
             return path;
         }
 
+        /// <summary>
+        /// Gets property for dynamic properties dictionary.
+        /// </summary>
+        /// <param name="openNode"></param>
+        /// <returns></returns>
```
Comment (mtroth-microsoft) [inline src/System.Web.OData/OData/Query/Expressions/ExpressionBinderBase.cs]: don't leave comments blank.

File: src/System.Web.OData/OData/Query/Expressions/ExpressionBinderBase.cs [hunk 7]
```diff
@@ -499,6 +500,31 @@ internal string GetFullPropertyPath(SingleValueNode node)
...
+        protected PropertyInfo GetDynamicPropertyContainer(SingleValueOpenPropertyAccessNode openNode)
+        {
+            IEdmStructuredType edmStructuredType;
+            var edmTypeReference = openNode.Source.TypeReference;
```
Comment (mtroth-microsoft) [inline src/System.Web.OData/OData/Query/Expressions/ExpressionBinderBase.cs]: >var [](start = 12, length = 3)

var

File: src/System.Web.OData/OData/Query/Expressions/ExpressionBinderBase.cs [hunk 8]
```diff
@@ -499,6 +500,31 @@ internal string GetFullPropertyPath(SingleValueNode node)
...
+            }
+            else if (edmTypeReference.IsComplex())
+            {
+                edmStructuredType = edmTypeReference.AsComplex().ComplexDefinition();
+            }
+            else
+            {
+                throw Error.NotSupported(SRResources.QueryNodeBindingNotSupported, openNode.Kind, typeof(FilterBinder).Name);
+            }
+            var prop = EdmLibHelpers.GetDynamicPropertyDictionary(edmStructuredType, Model);
```
Comment (mtroth-microsoft) [inline src/System.Web.OData/OData/Query/Expressions/ExpressionBinderBase.cs]: >var [](start = 12, length = 3)

var

File: src/System.Web.OData/OData/Query/Expressions/AggregationBinder.cs (hunk 4 above)
Comment (kosinsky) [reply src/System.Web.OData/OData/Query/Expressions/AggregationBinder.cs]: Are we going to get rid of vars in tests too?

Comment (mtroth-microsoft) [reply src/System.Web.OData/OData/Query/Expressions/AggregationBinder.cs]: it is lower priority, i think.

---
In reply to: [138944389](https://github.com/OData/WebApi/pull/1004#discussion_r138944389) [](ancestors = 138944389)
